src/bench/publisher.rs

```diff
@@ -214,38 +216,27 @@ impl Publisher {
 }
 
 /// make count number of requests at specified QoS.
-async fn requests(
-    topic: String,
-    count: usize,
-    client: AsyncClient,
-    qos: QoS,
-    delay: u64,
-    data_type: DataType,
-) {
-    let mut interval = match delay {
-        0 => None,
-        delay => Some(time::interval(time::Duration::from_millis(delay))),
-    };
-
-    for i in 0..count {
-        let payload = generate_data(i, data_type);
-        if let Some(interval) = &mut interval {
-            interval.tick().await;
-        }
+async fn requests(topic: String, client: AsyncClient, qos: QoS, generator: GenData) {
+    let mut stream = generator.into_stream();
+    let mut topic_cache: HashMap<DataEvent, String> = HashMap::new();
 
+    while let Some((event, payload)) = stream.next().await {
         // These errors are usually due to eventloop task being dead. We can ignore the
         // error here as the failed eventloop task would have already printed an error
-        if let Err(_e) = client.publish(topic.as_str(), qos, false, payload).await {
+        topic_cache
+            .entry(event)
+            .or_insert(topic.replace("{data_type}", &event.to_string()));
+        if let Err(_e) = client
+            .publish(topic_cache.get(&event).unwrap(), qos, false, payload)
+            .await
+        {
             break;
         }
-
-        info!("published {}", i);
     }
 
     if qos == QoS::AtMostOnce {
-        let payload = generate_data(count, data_type);
         if let Err(_e) = client
-            .publish(topic.as_str(), QoS::AtLeastOnce, false, payload)
+            .publish("test/complete", QoS::AtLeastOnce, false, "")
```

Comment:
we only use this for syncronization, and to get a `PubAck` back from server. We don't count this publish in subscriber because while loop above already send `count` number of publishes. Previously we where using this publish as well by only sending `count - 1` publish in the loop above.

This means when using `QoS::AtMostOnce` we will send an extra publish, but overhead of this will be negligible when count is high enough.